src/Plot/RobustPlot.purs

```diff
@@ -43,29 +45,109 @@ plotEnclosures canvasSize bounds domainSegments evaluator = segmentEnclosures
   toRange lower upper = Tuple lower upper
 
   toCanvasEnclosure :: Approx -> Maybe Polygon
-  toCanvasEnclosure x = case evaluator.f x of
-    Nothing -> Nothing
-    Just approxValue -> case evaluator.f xMidPoint of
-      Nothing -> Nothing
-      Just midApproxValue -> case evaluator.f' x of
-        Nothing -> Nothing
-        Just approxGradient -> Just polygon
+  {- overview:
+      - try to compute f'(x)
+        - if f''(x) is positive or negative, get f'(x) by endpoints
+        - otherwise get f'(x) by midpoint and f''(x), or directly if f''(x) fails to compute
+      - if f'(x) cannot be computed, use a box with f(x)
+      - if f'(x) is available, use it to compute parallelogram and intersect it with the f(x) box
+    -}
+  toCanvasEnclosure x =
+    let
+      xMidPointValue = map boundsA $ evaluator.f xMidPoint
+
+      xGradGrad = map boundsA $ evaluator.f'' x
+
+      xGradient = case xGradGrad of
+        Just (Tuple xGradGradLower xGradGradUpper)
+          | xGradGradLower !>=! zero -> do
+            xGradLeft <- evaluator.f' xLA
+            xGradRight <- evaluator.f' xUA
+            Just (Tuple (lowerA xGradLeft) (upperA xGradRight))
+          | xGradGradUpper !<=! zero -> do
+            xGradLeft <- evaluator.f' xLA
+            xGradRight <- evaluator.f' xUA
+            Just (Tuple (lowerA xGradRight) (upperA xGradLeft))
+          | otherwise -> map boundsA $ evaluator.f' x
+        _ -> map boundsA $ evaluator.f' x
+
+      xValue = case xGradient of
+        Just (Tuple xGradLower xGradUpper)
+          | xGradLower !>=! zero -> do
+            xLeft <- evaluator.f xLA
+            xRight <- evaluator.f xUA
+            Just (Tuple (lowerA xLeft) (upperA xRight))
+          | xGradUpper !<=! zero -> do
+            xLeft <- evaluator.f xLA
+            xRight <- evaluator.f xUA
+            Just (Tuple (lowerA xRight) (upperA xLeft))
+          | otherwise -> map boundsA $ evaluator.f x
+        _ -> map boundsA $ evaluator.f x
+    in
+      -- TODO: computer yLower yUpper by endpoints if gradient is positive or negative
```

Comment:
Is this TODO complete?

Reply:
Yes, I have just removed the comment.  Thank you.